agent/src/main/java/com/appland/appmap/process/hooks/Proxy.java

```diff
@@ -0,0 +1,155 @@
+
+    HookSite[] hookSites = allHookSites.get(method);
+    if (hookSites != null) {
+      invokeExcHook(hookSites, proxy, exc, methodArgs);
+    }
+  }
+
+  private static HookSite[] initHook(Method method, Object[] methodArgs) {
+    // Only hook methods that are matched by the config.
+    FullyQualifiedName fqn = new FullyQualifiedName(method);
+    if (AppMapConfig.get().includes(fqn) == null) {
+      return null;
+    }
+
+    AppMapClassPool.acquire(Thread.currentThread().getContextClassLoader());
+    try {
+      String[] argTypes = Arrays.stream(method.getParameterTypes()).map(a -> a.getName()).toArray(String[]::new);
+      CtBehavior behavior = ClassUtil.getDeclaredMethod(method.getDeclaringClass().getName(), method.getName(),
+          argTypes);
+      logger.trace("method: '{},'{}' behavior: {}", method::toString, method::toGenericString, behavior::getLongName);
+      List<Hook> hooks = ClassFileTransformer.getHooks(behavior.getName());
+      final List<HookSite> hookSites = hooks
+          .stream()
+          .map(hook -> hook.prepare(behavior))
+          .filter(Objects::nonNull)
+          .collect(Collectors.toList());
+
+      logger.trace("hookSites.size(): {}", hookSites.size());
+      if (hookSites.size() == 0) {
+        // not hooked
+        return null;
+      }
+      else {
+        logger.debug("{} hooks for method: '{},'{}' behavior: {}", hookSites::size, method::toString,
+            method::toGenericString, behavior::getLongName);
+      }
+      HookSite[] ret = new HookSite[3];
+      hookSites.stream().forEach(s -> ret[s.getMethodEvent().getIndex()] = s);
+      allHookSites.put(method, ret);
+
+      return ret;
+    } catch (NotFoundException e) {
+      logger.warn(e);
+    }
+    finally {
+      AppMapClassPool.release();
+    }
+
+    return null;
+  }
+
+  private static void invokeCallHook(HookSite[] hookSites, Object proxy, Method method, Object[] methodArgs) {
+    HookSite hookSite = hookSites[MethodEvent.METHOD_INVOCATION.getIndex()];
+    logger.trace("hookSite: {}, method event: {}, method: {}", hookSite::toString, hookSite::getMethodEvent,
+        method::getName);
+    Event event = EventTemplateRegistry.get().buildCallEvent(hookSite.getBehaviorOrdinal());
+    MethodCall.handle(event, proxy, methodArgs);
+  }
+
+  private static void invokeReturnHook(HookSite[] hookSites, Object proxy, Object ret, Object[] methodArgs) {
+    HookSite hookSite = hookSites[MethodEvent.METHOD_RETURN.getIndex()];
+    logger.trace("hookSite: {}, method event: {}, ret: {}", hookSite::toString, hookSite::getMethodEvent,
+        ret::toString);
+    Event event = EventTemplateRegistry.get().buildReturnEvent(hookSite.getBehaviorOrdinal());
+    MethodReturn.handle(event, proxy, ret, methodArgs);
+  }
+
+  private static void invokeExcHook(HookSite[] hookSites, Object proxy, Throwable exc, Object[] methodArgs) {
+    HookSite hookSite = hookSites[MethodEvent.METHOD_EXCEPTION.getIndex()];
```

Comment:
Is it possible for a hook to be missing? This isn't checked anywhere, I wonder if it should.